### PR TITLE
Add broker lifecycle listener

### PR DESCRIPTION
**Pull request by sirojan24**
Expose an OSGi service to listen broker lifecycle events.

### CODE REVIEW

File: components/andes/org.wso2.carbon.andes/src/main/java/org/wso2/carbon/andes/internal/QpidServiceDataHolder.java [hunk 1]
```diff
@@ -32,6 +36,8 @@
     private String accessKey = null;
     private ServerConfigurationService carbonConfiguration = null;
     private EventBundleNotificationService eventBundleNotificationService;
+    private List<BrokerShutdownListener>
+            brokerShutdownListeners = new ArrayList<>();
```
Comment (hemikak) [inline components/andes/org.wso2.carbon.andes/src/main/java/org/wso2/carbon/andes/internal/QpidServiceDataHolder.java]: I think we can bring this to the same line. The right margin should 120 characters or less.

File: components/andes/org.wso2.carbon.andes/src/main/java/org/wso2/carbon/andes/internal/QpidServiceDataHolder.java [hunk 2]
```diff
@@ -107,4 +113,23 @@ public void registerEventBundleNotificationService(
...
+        return brokerShutdownListeners;
+    }
+
+    /**
+     * Set the list of external andes server shutdown listeners
+     *
+     * @param brokerShutdownListeners
+     */
+    public void setBrokerShutdownListeners
+            (List<BrokerShutdownListener> brokerShutdownListeners) {
```
Comment (hemikak) [inline components/andes/org.wso2.carbon.andes/src/main/java/org/wso2/carbon/andes/internal/QpidServiceDataHolder.java]: I think we can bring this to the same line. The right margin should 120 characters or less.

File: components/andes/org.wso2.carbon.andes/src/main/java/org/wso2/carbon/andes/service/BrokerShutdownNotificationServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+ */
+public class BrokerShutdownNotificationServiceImpl implements
+        BrokerShutdownNotificationService {
+
+    /**
+     * register andes server shutdown listener
+     *
+     * @param brokerShutdownListener
+     */
+    @Override
```
Comment (hemikak) [inline components/andes/org.wso2.carbon.andes/src/main/java/org/wso2/carbon/andes/service/BrokerShutdownNotificationServiceImpl.java]: You can use the comment as follows : 
/**
 * {@inheritDoc}
*/

Comment (sirojan24) [inline components/andes/org.wso2.carbon.andes/src/main/java/org/wso2/carbon/andes/service/BrokerShutdownNotificationServiceImpl.java]: I'll use it

**Comment by a5anka** [comment]
Can we squash the two commits to a single one?

**Comment by sirojan24** [comment]
I squashed those into a single one

File: components/andes/org.wso2.carbon.andes/src/main/java/org/wso2/carbon/andes/internal/QpidServiceComponent.java [hunk 4]
```diff
@@ -171,7 +172,17 @@ protected void activate(ComponentContext context) throws AndesException {
 
                 public void startingShutdown() {
                     try {
+                        //executing pre-shutdown work for registered listeners before shutting down the andes server
+                        for(BrokerLifecycleListener listener: QpidServiceDataHolder.getInstance()
```
Comment (lahirus) [inline components/andes/org.wso2.carbon.andes/src/main/java/org/wso2/carbon/andes/internal/QpidServiceComponent.java]: formattin issue.

Comment (sirojan24) [inline components/andes/org.wso2.carbon.andes/src/main/java/org/wso2/carbon/andes/internal/QpidServiceComponent.java]: Removed in later PR

**Pull request merged**